File: core/shared/src/test/scala/diplomacy/variant/standard/rule/RuleSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.rule.OrderResult
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map.Power
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board.Keywords._
import diplomacy.variant.standard.board.Implicits._
import diplomacy.variant.standard.rule.Keywords._
import diplomacy.variant.standard.rule.Implicits._

class RuleSpec extends UnitSpec {
  val rule = new Rule[T, map.Power](T.nextTurn)
  type Executed =
    OrderResult.Executed[map.Power, rule.MilitaryBranch, rule.Order, rule.Result]

  "A Rule" when {
    "there are no dislodged units" should {
      "skip retreat phase." in {
        val board =
          new Board[T, map.Power](
            map.map,
            1901.Spring - Movement,
            Set(
              DiplomacyUnit(France, Army, Bre),
              DiplomacyUnit(Germany, Army, Pic)
            ), Map(), Map()
          )
        val ruleHelper = StandardRuleOrderHelper(board)
        import ruleHelper._

        val Right(result) = rule.resolve(board, Set(A(Bre).hold(), A(Pic).hold()))
        result.board should be (board.copy(state = 1901.Autumn - Movement))
      }
    }
    "all dislodged units cannot retreat" should {
      "skip retreat phase." in {
        val board =
          new Board[T, map.Power](
            map.map,
            1901.Spring - Movement,
            Set(
              DiplomacyUnit(France, Army, Bre),
              DiplomacyUnit(Germany, Army, Gas),
              DiplomacyUnit(Germany, Army, Par),
              DiplomacyUnit(Germany, Army, Pic)
            ), Map(), Map()
          )
        val ruleHelper = StandardRuleOrderHelper(board)
        import ruleHelper._

        val Right(result) = rule.resolve(board, Set(
          A(Bre).hold(),
          A(Pic).move(Bre), A(Par).support(A(Pic).move(Bre)), A(Gas).support(A(Pic).move(Bre))
        ))
        result.board should be (board.copy(
          state = 1901.Autumn - Movement,
          units = Set(
              DiplomacyUnit(Germany, Army, Gas),
              DiplomacyUnit(Germany, Army, Par),
              DiplomacyUnit(Germany, Army, Bre)
            )
        ))
        result.result should be(Set(
          new Executed(A(Bre).hold(), Result.Dislodged),
          new Executed(A(Bre).disband(), Result.Success),
          new Executed(A(Pic).move(Bre), Result.Success),
          new Executed(A(Par).support(A(Pic).move(Bre)), Result.Success),
          new Executed(A(Gas).support(A(Pic).move(Bre)), Result.Success)
        ))
      }
    }
    "there are no power that can build units" should {
      "skip build phase." in {
        val board =
          new Board[T, map.Power](
            map.map,
            1901.Autumn - Movement,
            Set(
              DiplomacyUnit(France, Army, Bre),
              DiplomacyUnit(Germany, Army, Mun)
            ),
            Map(),
            Map(
              Bre.province -> ProvinceStatus[map.Power](Option(France), false),
              Mun.province -> ProvinceStatus[map.Power](Option(Germany), false)
            )
          )
        val ruleHelper = StandardRuleOrderHelper(board)
        import ruleHelper._

        val Right(result) = rule.resolve(board, Set(A(Bre).hold(), A(Mun).hold()))
        result.board should be (board.copy(state = 1902.Spring - Movement))
      }
    }
    "all build and disband orders are automatically decidable" should {
      "skip build phase." in {
        val board =
          new Board[T, map.Power](
            map.map,
            1901.Autumn - Movement,
            Set(
              DiplomacyUnit(France, Army, Gas),
              DiplomacyUnit(Germany, Army, Mun)
            ),
            Map(),
            Map(
              Gas.province -> ProvinceStatus[map.Power](Option(France), false),
              Mun.province -> ProvinceStatus[map.Power](Option(Germany), false)
            )
          )
        val ruleHelper = StandardRuleOrderHelper(board)
        import ruleHelper._

        val Right(result) = rule.resolve(board, Set(A(Gas).hold(), A(Mun).hold()))
        result.board should be (board.copy(
          state = 1902.Spring - Movement,
          units = Set(DiplomacyUnit(Germany, Army, Mun))
        ))
        result.result should be(Set(
          new Executed(A(Gas).hold(), Result.Success),
          new Executed(A(Gas).disband(), Result.Success),
          new Executed(A(Mun).hold(), Result.Success)
        ))
      }
    }
    "a power controls half of supply centers" should {
      "finish the game." in {
        val board =
          new Board[T, map.Power](
            map.map, 1901.Autumn - Movement,
            Set(),
            Map(),
            Map(
              StP.province -> ProvinceStatus[Power](Option(France), false),
              Swe.province -> ProvinceStatus[Power](Option(France), false),
              Nwy.province -> ProvinceStatus[Power](Option(France), false),
              Den.province -> ProvinceStatus[Power](Option(France), false),
              Lvp.province -> ProvinceStatus[Power](Option(France), false),
              Edi.province -> ProvinceStatus[Power](Option(France), false),
              Lon.province -> ProvinceStatus[Power](Option(France), false),
              Ber.province -> ProvinceStatus[Power](Option(France), false),
              Kie.province -> ProvinceStatus[Power](Option(France), false),
              Mun.province -> ProvinceStatus[Power](Option(France), false),
              Hol.province -> ProvinceStatus[Power](Option(France), false),
              Bel.province -> ProvinceStatus[Power](Option(France), false),
              Bre.province -> ProvinceStatus[Power](Option(France), false),
              Par.province -> ProvinceStatus[Power](Option(France), false),
              Mar.province -> ProvinceStatus[Power](Option(France), false),
              Spa.province -> ProvinceStatus[Power](Option(France), false),
              Por.province -> ProvinceStatus[Power](Option(France), false),
              Tun.province -> ProvinceStatus[Power](Option(France), false)
            )
          )
        val $ = StandardRuleOrderHelper(board)
        import $._

        val Right(result) = rule.resolve(board, Set())
        result.isFinished should be(true)
      }
    }
  }
}

File: test/variant/standard/rule/rule-spec.ts
import * as chai from "chai"
import {
  Rule, Board, Unit, MilitaryBranch, State, Result, Phase, ProvinceStatus, Dislodged,
  Disband, StandardRuleHelper as Helper
} from "../../../../src/variant/standard/rule/module"
import { Executed } from "./../../../../src/rule/module"
import { locations as $, Power, map } from "./../../../../src/variant/standard/map/module"
import { Season, Turn } from "./../../../../src/variant/standard/board/module"

const { Army, Fleet } = MilitaryBranch
const { Retreat, Movement } = Phase
const { Spring, Autumn } = Season

const should = chai.should()

const rule = new Rule((x: Power) => Phase[x])

const state1 = new State(new Turn(1901, Spring), Movement)
const state3 = new State(new Turn(1901, Autumn), Movement)
const state4 = new State(new Turn(1901, Autumn), Retreat)
const state5 = new State(new Turn(1902, Spring), Movement)

describe("A rule", () => {
  describe("when there are no dislodged units", () => {
    it("skips retreat phase.", () => {
      const b = new Board<Power>(
        map, state1,
        [new Unit(Army, $.Bre, Power.France), new Unit(Army, $.Pic, Power.Germany)],
        [], []
      )
      const $$ = new Helper(b)
      const r = rule.resolve(b, new Set()).result
      should.not.equal(r, null)
      if (r) {
        const { board, results } = r

        Array.from(results).should.have.deep.members([
          new Executed($$.A($.Bre).hold(), Result.Success),
          new Executed($$.A($.Pic).hold(), Result.Success)
        ])
        board.should.deep.equal(new Board<Power>(
          map, state3,
          [new Unit(Army, $.Bre, Power.France), new Unit(Army, $.Pic, Power.Germany)],
          [], []
        ))
      }
    })
  })
  describe("when all dislodged units cannot retreat", () => {
    it("skips retreat phase", () => {
      const b = new Board<Power>(
        map, state1,
        [
          new Unit(Army, $.Bre, Power.France), new Unit(Army, $.Gas, Power.Germany),
          new Unit(Army, $.Par, Power.Germany), new Unit(Army, $.Pic, Power.Germany)
        ],
        [], []
      )
      const $$ = new Helper(b)
      const r = rule.resolve(
        b,
        new Set([
          $$.A($.Bre).hold(), $$.A($.Pic).move($.Bre),
          $$.A($.Par).support($$.A($.Pic).move($.Bre)), $$.A($.Gas).support($$.A($.Pic).move($.Bre))
        ])
      ).result
      should.not.equal(r, null)
      if (r) {
        const { board, results } = r

        Array.from(results).should.have.deep.members([
          new Executed($$.A($.Bre).hold(), Result.Dislodged),
          new Executed($$.A($.Pic).move($.Bre), Result.Success),
          new Executed($$.A($.Par).support($$.A($.Pic).move($.Bre)), Result.Success),
          new Executed($$.A($.Gas).support($$.A($.Pic).move($.Bre)), Result.Success),
          new Executed($$.A($.Bre).disband(), Result.Success)
        ])
        board.should.deep.equal(new Board<Power>(
          map, state3,
          [
            new Unit(Army, $.Gas, Power.Germany),
            new Unit(Army, $.Par, Power.Germany), new Unit(Army, $.Bre, Power.Germany)
          ],
          [], []
        ))
      }
    })
  })
  describe("when there are no force that can build units", () => {
    it("skips build phase.", () => {
      const b = new Board<Power>(
        map, state4,
        [
          new Unit(Army, $.Bre, Power.France), new Unit(Army, $.Mun, Power.Germany)
        ],
        [],
        [
          [$.Bre.province, new ProvinceStatus(Power.France, false)],
          [$.Mun.province, new ProvinceStatus(Power.Germany, false)]
        ]
      )
      const r = rule.resolve(b, new Set()).result
      should.not.equal(r, null)
      if (r) {
        const { board, results } = r

        results.should.deep.equal(new Set())
        board.should.deep.equal(new Board<Power>(
          map, state5,
          [
            new Unit(Army, $.Bre, Power.France), new Unit(Army, $.Mun, Power.Germany)
          ],
          [],
          [
            [$.Bre.province, new ProvinceStatus(Power.France, false)],
            [$.Mun.province, new ProvinceStatus(Power.Germany, false)]
          ]
        ))
      }
    })
  })
  describe("when all build or disband orders are automatically decidable", () => {
    it("skip build phase.", () => {
      const b = new Board<Power>(
        map, state3,
        [
          new Unit(Army, $.Gas, Power.France), new Unit(Army, $.Mun, Power.Germany)
        ],
        [],
        [
          [$.Gas.province, new ProvinceStatus(Power.France, false)],
          [$.Mun.province, new ProvinceStatus(Power.Germany, false)]
        ]
      )
      const $$ = new Helper(b)
      const r = rule.resolve(b, new Set()).result
      should.not.equal(r, null)

      if (r) {
        const { board, results } = r

        Array.from(results).should.have.deep.members([
          new Executed($$.A($.Gas).hold(), Result.Success),
          new Executed($$.A($.Mun).hold(), Result.Success),
          new Executed(new Disband(new Unit(Army, $.Gas, Power.France)), Result.Success)
        ])
        board.should.deep.equal(new Board<Power>(
          map, state5,
          [new Unit(Army, $.Mun, Power.Germany)],
          [],
          [
            [$.Gas.province, new ProvinceStatus(Power.France, false)],
            [$.Mun.province, new ProvinceStatus(Power.Germany, false)]
          ]
        ))
      }
    })
  })
  describe("when a power controls half of supply centers", () => {
    it("finishes the game", () => {
      const b = new Board<Power>(
        map, state3,
        [],
        [],
        [
          [$.StP.province, new ProvinceStatus(Power.France, false)],
          [$.Swe.province, new ProvinceStatus(Power.France, false)],
          [$.Nwy.province, new ProvinceStatus(Power.France, false)],
          [$.Den.province, new ProvinceStatus(Power.France, false)],
          [$.Lvp.province, new ProvinceStatus(Power.France, false)],
          [$.Edi.province, new ProvinceStatus(Power.France, false)],
          [$.Lon.province, new ProvinceStatus(Power.France, false)],
          [$.Ber.province, new ProvinceStatus(Power.France, false)],
          [$.Kie.province, new ProvinceStatus(Power.France, false)],
          [$.Mun.province, new ProvinceStatus(Power.France, false)],
          [$.Hol.province, new ProvinceStatus(Power.France, false)],
          [$.Bel.province, new ProvinceStatus(Power.France, false)],
          [$.Bre.province, new ProvinceStatus(Power.France, false)],
          [$.Par.province, new ProvinceStatus(Power.France, false)],
          [$.Mar.province, new ProvinceStatus(Power.France, false)],
          [$.Spa.province, new ProvinceStatus(Power.France, false)],
          [$.Por.province, new ProvinceStatus(Power.France, false)],
          [$.Tun.province, new ProvinceStatus(Power.France, false)]
        ]
      )
      const r = rule.resolve(b, new Set()).result
      should.not.equal(r, null)
      if (r) {
        r.isFinished.should.equal(true)
      }
    })
  })
})
